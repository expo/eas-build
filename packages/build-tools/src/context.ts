import path from 'path';

import fs from 'fs-extra';
import {
  ManagedArtifactType,
  BuildPhase,
  BuildPhaseResult,
  BuildPhaseStats,
  Job,
  LogMarker,
  Env,
  errors,
  Metadata,
  EnvironmentSecretType,
  GenericArtifactType,
  isGenericArtifact,
} from '@expo/eas-build-job';
import { ExpoConfig } from '@expo/config';
import { bunyan } from '@expo/logger';
import { SpawnPromise, SpawnOptions, SpawnResult } from '@expo/turtle-spawn';
import { BuildTrigger } from '@expo/eas-build-job/dist/common';

import { PackageManager, resolvePackageManager } from './utils/packageManager';
import { resolveBuildPhaseErrorAsync } from './buildErrors/detectError';
import { readAppConfig } from './utils/appConfig';
import { createTemporaryEnvironmentSecretFile } from './utils/environmentSecrets';

export type Artifacts = Partial<Record<ManagedArtifactType, string>>;

export interface CacheManager {
  saveCache(ctx: BuildContext<Job>): Promise<void>;
  restoreCache(ctx: BuildContext<Job>): Promise<void>;
}

export interface LogBuffer {
  getLogs(): string[];
  getPhaseLogs(buildPhase: string): string[];
}

export type ArtifactToUpload =
  | {
      type: ManagedArtifactType;
      paths: string[];
    }
  | {
      type: GenericArtifactType;
      key: string;
      paths: string[];
    };

export interface BuildContextOptions {
  workingdir: string;
  logger: bunyan;
  logBuffer: LogBuffer;
  env: Env;
  cacheManager?: CacheManager;
  /**
   * @deprecated
   */
  runGlobalExpoCliCommand: (
    args: string[],
    options: SpawnOptions,
    npmVersionAtLeast7: boolean
  ) => SpawnPromise<SpawnResult>;
  uploadArtifact: (spec: { artifact: ArtifactToUpload; logger: bunyan }) => Promise<string | null>;
  reportError?: (
    msg: string,
    err?: Error,
    options?: { tags?: Record<string, string>; extras?: Record<string, string> }
  ) => void;
  reportBuildPhaseStats?: (stats: BuildPhaseStats) => void;
  skipNativeBuild?: boolean;
  metadata?: Metadata;
}

export class SkipNativeBuildError extends Error {}

export class BuildContext<TJob extends Job> {
  public readonly workingdir: string;
  public logger: bunyan;
  public readonly logBuffer: LogBuffer;
  public readonly cacheManager?: CacheManager;
  /**
   * @deprecated
   */
  public readonly runGlobalExpoCliCommand: (
    args: string[],
    options: SpawnOptions,
    npmVersionAtLeast7: boolean
  ) => SpawnPromise<SpawnResult>;
  public readonly reportError?: (
    msg: string,
    err?: Error,
    options?: { tags?: Record<string, string>; extras?: Record<string, string> }
  ) => void;
  public readonly skipNativeBuild?: boolean;
  public artifacts: Artifacts = {};

  private _env: Env;
  private _job: TJob;
  private _metadata?: Metadata;
  private readonly defaultLogger: bunyan;
  private readonly _uploadArtifact: BuildContextOptions['uploadArtifact'];
  private buildPhase?: BuildPhase;
  private buildPhaseSkipped = false;
  private buildPhaseHasWarnings = false;
  private _appConfig?: ExpoConfig;
  private readonly reportBuildPhaseStats?: (stats: BuildPhaseStats) => void;

  constructor(job: TJob, options: BuildContextOptions) {
    this.workingdir = options.workingdir;
    this.defaultLogger = options.logger;
    this.logger = this.defaultLogger;
    this.logBuffer = options.logBuffer;
    this.cacheManager = options.cacheManager;
    this.runGlobalExpoCliCommand = options.runGlobalExpoCliCommand;
    this._uploadArtifact = options.uploadArtifact;
    this.reportError = options.reportError;
    this._job = job;
    this._metadata = options.metadata;
    this.skipNativeBuild = options.skipNativeBuild;
    this.reportBuildPhaseStats = options.reportBuildPhaseStats;

    const environmentSecrets = this.getEnvironmentSecrets(job);
    this._env = {
      ...options.env,
      ...job?.builderEnvironment?.env,
      ...environmentSecrets,
      __EAS_BUILD_ENVS_DIR: this.buildEnvsDirectory,
    };
    this._env.PATH = this._env.PATH
      ? [this.buildExecutablesDirectory, this._env.PATH].join(':')
      : this.buildExecutablesDirectory;
  }

  public get job(): TJob {
    return this._job;
  }
  public get metadata(): Metadata | undefined {
    return this._metadata;
  }
  public get env(): Env {
    return this._env;
  }
  public get buildDirectory(): string {
    return path.join(this.workingdir, 'build');
  }
  public get buildLogsDirectory(): string {
    return path.join(this.workingdir, 'logs');
  }
  /**
   * Directory used to store executables used during regular (non-custom) builds.
   */
  public get buildExecutablesDirectory(): string {
    return path.join(this.workingdir, 'bin');
  }
  /**
   * Directory used to store env variables registered in the current build step.
   * All values stored here will be available in the next build phase as env variables.
   */
  public get buildEnvsDirectory(): string {
    return path.join(this.workingdir, 'env');
  }
  public get environmentSecretsDirectory(): string {
    return path.join(this.workingdir, 'environment-secrets');
  }
  public get packageManager(): PackageManager {
    return resolvePackageManager(this.getReactNativeProjectDirectory());
  }
  public get appConfig(): ExpoConfig {
    if (!this._appConfig) {
      this._appConfig = readAppConfig({
        projectDir: this.getReactNativeProjectDirectory(),
        env: this.env,
        logger: this.logger,
        sdkVersion: this.metadata?.sdkVersion,
      }).exp;
    }
    return this._appConfig;
  }

  public async runBuildPhase<T>(
    buildPhase: BuildPhase,
    phase: () => Promise<T>,
    {
      doNotMarkStart = false,
      doNotMarkEnd = false,
    }: {
      doNotMarkStart?: boolean;
      doNotMarkEnd?: boolean;
    } = {}
  ): Promise<T> {
    let startTimestamp = Date.now();
    try {
      this.setBuildPhase(buildPhase, { doNotMarkStart });
      startTimestamp = Date.now();
      const result = await phase();
      const durationMs = Date.now() - startTimestamp;
      const buildPhaseResult: BuildPhaseResult = this.buildPhaseSkipped
        ? BuildPhaseResult.SKIPPED
        : this.buildPhaseHasWarnings
          ? BuildPhaseResult.WARNING
          : BuildPhaseResult.SUCCESS;
      await this.endCurrentBuildPhaseAsync({ result: buildPhaseResult, doNotMarkEnd, durationMs });
      return result;
    } catch (err: any) {
      const durationMs = Date.now() - startTimestamp;
      const resolvedError = await this.handleBuildPhaseErrorAsync(err, buildPhase);
      await this.endCurrentBuildPhaseAsync({ result: BuildPhaseResult.FAIL, durationMs });
      throw resolvedError;
    }
  }

  public markBuildPhaseSkipped(): void {
    this.buildPhaseSkipped = true;
  }

  public markBuildPhaseHasWarnings(): void {
    this.buildPhaseHasWarnings = true;
  }

  public async uploadArtifact({
    artifact,
    logger,
  }: {
    artifact: ArtifactToUpload;
    logger: bunyan;
  }): Promise<void> {
    const bucketKey = await this._uploadArtifact({ artifact, logger });
    if (bucketKey && !isGenericArtifact(artifact)) {
      this.artifacts[artifact.type] = bucketKey;
    }
  }

  public updateEnv(env: Env): void {
    if (this._job.triggeredBy !== BuildTrigger.GIT_BASED_INTEGRATION) {
      throw new Error(
        'Updating environment variables is only allowed when build was triggered by a git-based integration.'
      );
    }
    this._env = {
      ...env,
      ...this._env,
      __EAS_BUILD_ENVS_DIR: this.buildEnvsDirectory,
    };
    this._env.PATH = this._env.PATH
      ? [this.buildExecutablesDirectory, this._env.PATH].join(':')
      : this.buildExecutablesDirectory;
  }

  public updateJobInformation(job: TJob, metadata: Metadata): void {
    if (this._job.triggeredBy !== BuildTrigger.GIT_BASED_INTEGRATION) {
      throw new Error(
        'Updating job information is only allowed when build was triggered by a git-based integration.'
      );
    }
    this._job = { ...job, triggeredBy: this._job.triggeredBy };
    this._metadata = metadata;
  }

  private async handleBuildPhaseErrorAsync(
    err: any,
    buildPhase: BuildPhase
  ): Promise<errors.BuildError> {
    const buildError = await resolveBuildPhaseErrorAsync(
      err,
      this.logBuffer.getPhaseLogs(buildPhase),
      {
        job: this.job,
        phase: buildPhase,
        env: this.env,
      },
      this.buildLogsDirectory
    );
    if (buildError.errorCode === errors.ErrorCode.UNKNOWN_ERROR) {
      // leaving message empty, website will display err.stack which already includes err.message
      this.logger.error({ err }, '');
    } else {
      this.logger.error(`Error: ${buildError.userFacingMessage}`);
    }
    return buildError;
  }

  public getReactNativeProjectDirectory(baseDirectory = this.buildDirectory): string {
    if (!this.job.platform) {
      // Generic jobs start from base directory.
      return baseDirectory;
    }

    return path.join(baseDirectory, this.job.projectRootDirectory ?? '.');
  }

  private setBuildPhase(buildPhase: BuildPhase, { doNotMarkStart = false } = {}): void {
    if (this.buildPhase) {
      if (this.buildPhase === buildPhase) {
        return;
      } else {
        this.logger.info(
          { marker: LogMarker.END_PHASE, result: BuildPhaseResult.UNKNOWN },
          `End phase: ${this.buildPhase}`
        );
        this.logger = this.defaultLogger;
      }
    }
    this.buildPhase = buildPhase;
    this.logger = this.defaultLogger.child({ phase: buildPhase });
    if (!doNotMarkStart) {
      this.logger.info({ marker: LogMarker.START_PHASE }, `Start phase: ${this.buildPhase}`);
    }
  }

  private async endCurrentBuildPhaseAsync({
    result,
    doNotMarkEnd = false,
    durationMs,
  }: {
    result: BuildPhaseResult;
    doNotMarkEnd?: boolean;
    durationMs: number;
  }): Promise<void> {
    if (!this.buildPhase) {
      return;
    }
    await this.collectAndUpdateEnvVariablesAsync();

    this.reportBuildPhaseStats?.({ buildPhase: this.buildPhase, result, durationMs });

    if (!doNotMarkEnd) {
      this.logger.info(
        { marker: LogMarker.END_PHASE, result, durationMs },
        `End phase: ${this.buildPhase}`
      );
    }
    this.logger = this.defaultLogger;
    this.buildPhase = undefined;
    this.buildPhaseSkipped = false;
    this.buildPhaseHasWarnings = false;
  }

  private async collectAndUpdateEnvVariablesAsync(): Promise<void> {
    const filenames = await fs.readdir(this.buildEnvsDirectory);

    const entries = await Promise.all(
      filenames.map(async (basename) => {
        const rawContents = await fs.readFile(
          path.join(this.buildEnvsDirectory, basename),
          'utf-8'
        );
        return [basename, rawContents];
      })
    );
    await Promise.all(
      filenames.map(async (basename) => {
        await fs.remove(path.join(this.buildEnvsDirectory, basename));
      })
    );
    this._env = {
      ...this._env,
      ...Object.fromEntries(entries),
    };
  }

  private getEnvironmentSecrets(job: TJob): Record<string, string> {
    if (!job?.secrets?.environmentSecrets) {
      return {};
    }

    const environmentSecrets: Record<string, string> = {};
    for (const { name, type, value } of job.secrets.environmentSecrets) {
      if (type === EnvironmentSecretType.STRING) {
        environmentSecrets[name] = value;
      } else {
        environmentSecrets[name] = createTemporaryEnvironmentSecretFile(
          this.environmentSecretsDirectory,
          value
        );
      }
    }
    return environmentSecrets;
  }
}
