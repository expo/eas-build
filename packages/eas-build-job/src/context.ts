import { DynamicInterpolationContext, Env, StaticWorkflowInterpolationContext } from './common';
import { Job } from './job';
import { Metadata } from './metadata';

type StaticJobOnlyInterpolationContext = {
  job: Job;
  metadata: Metadata | null;
  env: Env;
  steps: Record<
    string,
    {
      outputs: Record<string, string | undefined>;
    }
  >;
};

export type StaticJobInterpolationContext =
  | (StaticWorkflowInterpolationContext & StaticJobOnlyInterpolationContext)
  | StaticJobOnlyInterpolationContext;

export type JobInterpolationContext = StaticJobInterpolationContext & DynamicInterpolationContext;
