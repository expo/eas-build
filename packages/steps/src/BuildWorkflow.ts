import { BuildFunctionById } from './BuildFunction.js';
import { BuildStep } from './BuildStep.js';
import { BuildStepGlobalContext } from './BuildStepContext.js';

export class BuildWorkflow {
  public readonly buildSteps: BuildStep[];
  public readonly buildFunctions: BuildFunctionById;

  constructor(
    // @ts-expect-error ctx is not used in this class but let's keep it here for consistency
    private readonly ctx: BuildStepGlobalContext,
    { buildSteps, buildFunctions }: { buildSteps: BuildStep[]; buildFunctions: BuildFunctionById }
  ) {
    this.buildSteps = buildSteps;
    this.buildFunctions = buildFunctions;
  }

  public async executeAsync(): Promise<void> {
    for (const step of this.buildSteps) {
      await step.executeAsync();
    }
  }
}
