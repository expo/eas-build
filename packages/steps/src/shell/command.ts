import os from 'os';

// stolen from https://circleci.com/docs/configuration-reference/
const DEFAULT_SHELL = '/bin/bash -eo pipefail';
const SHELL_BY_PLATFORM: Partial<Record<NodeJS.Platform, string>> = {
  darwin: '/bin/bash --login -eo pipefail',
};

export function getDefaultShell(platform: NodeJS.Platform = os.platform()): string {
  const platformSpecificShell = SHELL_BY_PLATFORM[platform];
  if (platformSpecificShell) {
    return platformSpecificShell;
  } else {
    return DEFAULT_SHELL;
  }
}

export function getShellCommandAndArgs(
  shell: string,
  script?: string
): { command: string; args?: string[] } {
  const splits = shell.split(' ');
  const command = splits[0];
  const args = [...splits.slice(1)];
  if (script) {
    args.push(script);
  }
  return {
    command,
    args,
  };
}
