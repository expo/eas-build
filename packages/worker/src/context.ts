import assert from 'assert';

import { BuildContext, BuildContextOptions, LogBuffer } from '@expo/build-tools';
import { ManagedArtifactType, BuildPhaseStats, Job, Metadata, Platform } from '@expo/eas-build-job';
import { bunyan } from '@expo/logger';

import { Analytics } from './external/analytics';
import config from './config';
import { GCSCacheManager } from './CacheManager';
import sentry from './sentry';
import {
  uploadApplicationArchiveAsync,
  uploadBuildArtifactsAsync,
  uploadWithAnalyticsAsync,
  uploadWorkflowArtifactAsync,
} from './upload';
import { uploadXcodeBuildLogs } from './ios/xcodeLogs';
import { getBuildEnv } from './env';

export function createBuildContext<TJob extends Job>({
  job,
  logBuffer,
  analytics,
  metadata,
  projectId,
  buildId,
  buildLogger,
  reportBuildPhaseStatsFn,
}: {
  job: TJob;
  logBuffer: LogBuffer;
  analytics: Analytics;
  metadata: Metadata;
  projectId: string;
  buildId: string;
  buildLogger: bunyan;
  reportBuildPhaseStatsFn: (stats: BuildPhaseStats) => void;
}): BuildContext<TJob> {
  const env = getBuildEnv({ job, projectId, metadata, buildId });
  const childLogger = buildLogger.child({ buildId });

  const uploadArtifact: BuildContextOptions['uploadArtifact'] = async ({ artifact, logger }) => {
    const { paths, type } = artifact;

    switch (type) {
      case ManagedArtifactType.XCODE_BUILD_LOGS: {
        if (job.platform === Platform.IOS) {
          await uploadXcodeBuildLogs(logger, paths[0]);
          return null;
        }
        throw new Error('Uploading Xcode logs in non-iOS builds is not supported');
      }
      case ManagedArtifactType.APPLICATION_ARCHIVE: {
        assert(job.platform, 'Uploading application archives outside of builds is not supported.');
        return await uploadWithAnalyticsAsync(
          () => uploadApplicationArchiveAsync(ctx, { artifactPaths: paths, buildId, logger }),
          analytics
        );
      }
      case ManagedArtifactType.BUILD_ARTIFACTS: {
        assert(job.platform, 'Uploading build artifacts outside of builds is not supported.');
        return await uploadWithAnalyticsAsync(
          () => uploadBuildArtifactsAsync(ctx, { artifactPaths: paths, buildId, logger }),
          analytics
        );
      }
      default: {
        await uploadWithAnalyticsAsync(
          () =>
            uploadWorkflowArtifactAsync(ctx, {
              artifactPaths: paths,
              logger,
              name: artifact.name,
            }),
          analytics
        );
        return null;
      }
    }
  };

  const ctx = new BuildContext<TJob>(job, {
    workingdir: config.workingdir,
    logger: childLogger,
    logBuffer,
    env,
    uploadArtifact,
    reportError: (msg, err, { tags, extras } = {}) => {
      sentry.handleError(msg, err, { tags, extras });
    },
    cacheManager: new GCSCacheManager(),
    metadata,
    reportBuildPhaseStats: reportBuildPhaseStatsFn,
  });
  return ctx;
}
