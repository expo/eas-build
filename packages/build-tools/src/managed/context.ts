import { Ios, Android } from '@expo/eas-build-job';

import { BuildContext, BuildContextOptions } from '../context';

import { NpxExpoCliEjectProvider } from './NpxExpoCliEject';
import { EjectProvider } from './EjectProvider';

export type ManagedJob = Ios.ManagedJob | Android.ManagedJob;

interface ManagedBuildContextOptions<TJob extends ManagedJob> extends BuildContextOptions {
  ejectProvider?: EjectProvider<TJob>;
}

export class ManagedBuildContext<TJob extends ManagedJob> extends BuildContext<TJob> {
  public readonly ejectProvider: EjectProvider<TJob>;

  constructor(job: TJob, { ejectProvider, ...otherOptions }: ManagedBuildContextOptions<TJob>) {
    super(job, otherOptions);
    this.ejectProvider = ejectProvider ?? new NpxExpoCliEjectProvider();
  }
}
