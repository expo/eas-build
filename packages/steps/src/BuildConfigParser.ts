import fs from 'fs';
import path from 'path';

import { v4 as uuidv4 } from 'uuid';
import YAML from 'yaml';

import {
  BuildStepConfig,
  BuildStepInputs,
  BuildStepOutputs,
  validateBuildConfig,
} from './BuildConfig.js';
import { BuildStep } from './BuildStep.js';
import { BuildStepContext } from './BuildStepContext.js';
import { BuildStepInput } from './BuildStepInput.js';
import { BuildStepOutput } from './BuildStepOutput.js';
import { BuildWorkflow } from './BuildWorkflow.js';

export class BuildConfigParser {
  private readonly configPath: string;

  constructor(private readonly ctx: BuildStepContext, { configPath }: { configPath: string }) {
    this.configPath = configPath;
  }

  public async parseAsync(): Promise<BuildWorkflow> {
    const rawConfig = await this.readRawConfigAsync();
    const config = validateBuildConfig(rawConfig);
    const steps = config.build.steps.map((stepConfig) =>
      this.createBuildStepFromConfig(stepConfig)
    );
    return new BuildWorkflow({ buildSteps: steps });
  }

  private async readRawConfigAsync(): Promise<any> {
    const contents = await fs.promises.readFile(this.configPath, 'utf-8');
    return YAML.parse(contents);
  }

  private createBuildStepFromConfig(buildStepConfig: BuildStepConfig): BuildStep {
    if (typeof buildStepConfig === 'string') {
      // TODO: implement calling functions
      throw new Error('Not implemented yet');
    } else if (typeof buildStepConfig.run === 'string') {
      const command = buildStepConfig.run;
      return new BuildStep(this.ctx, {
        id: uuidv4(),
        workingDirectory: this.ctx.workingDirectory,
        command,
      });
    } else {
      const {
        id,
        inputs: inputsConfig,
        outputs: outputsConfig,
        name,
        workingDirectory,
        shell,
        command,
      } = buildStepConfig.run;
      const inputs = inputsConfig && this.createBuildStepInputsFromConfig(inputsConfig);
      const outputs = outputsConfig && this.createBuildStepOutputsFromConfig(outputsConfig);
      return new BuildStep(this.ctx, {
        id: id ?? uuidv4(),
        inputs,
        outputs,
        name,
        workingDirectory:
          workingDirectory !== undefined
            ? path.resolve(this.ctx.workingDirectory, workingDirectory)
            : this.ctx.workingDirectory,
        shell,
        command,
      });
    }
  }

  private createBuildStepInputsFromConfig(
    buildStepInputsConfig: BuildStepInputs
  ): BuildStepInput[] {
    const inputs: BuildStepInput[] = [];
    for (const [key, value] of Object.entries(buildStepInputsConfig)) {
      const input = new BuildStepInput(this.ctx, { id: key, defaultValue: value, required: true });
      inputs.push(input);
    }
    return inputs;
  }

  private createBuildStepOutputsFromConfig(
    buildStepOutputsConfig: BuildStepOutputs
  ): BuildStepOutput[] {
    const outputs: BuildStepOutput[] = [];
    for (const entry of buildStepOutputsConfig) {
      const output =
        typeof entry === 'string'
          ? new BuildStepOutput(this.ctx, { id: entry, required: true })
          : new BuildStepOutput(this.ctx, { id: entry.name, required: entry.required ?? true });
      outputs.push(output);
    }
    return outputs;
  }
}
