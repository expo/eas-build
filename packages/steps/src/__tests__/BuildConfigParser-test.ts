import path from 'path';
import url from 'url';

import { BuildConfigParser } from '../BuildConfigParser.js';
import { BuildWorkflow } from '../BuildWorkflow.js';
import { getDefaultShell } from '../utils/shell/command.js';

import { createMockContext } from './utils/context.js';

const UUID_REGEX =
  /^[0-9a-fA-F]{8}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{12}$/;

const __dirname = url.fileURLToPath(new URL('.', import.meta.url));

describe(BuildConfigParser, () => {
  describe(BuildConfigParser.prototype.parseAsync, () => {
    it('returns a BuildWorkflow object', async () => {
      const ctx = createMockContext();
      const parser = new BuildConfigParser(ctx, {
        configPath: path.join(__dirname, './fixtures/build.yml'),
      });
      const result = await parser.parseAsync();
      expect(result).toBeInstanceOf(BuildWorkflow);
    });

    it('parses steps from the build workflow', async () => {
      const ctx = createMockContext();
      const parser = new BuildConfigParser(ctx, {
        configPath: path.join(__dirname, './fixtures/build.yml'),
      });
      const workflow = await parser.parseAsync();
      const buildSteps = workflow.buildSteps;
      expect(buildSteps.length).toBe(6);

      // - run: echo "Hi!"
      const step1 = buildSteps[0];
      expect(step1.id).toMatch(UUID_REGEX);
      expect(step1.name).toBeUndefined();
      expect(step1.command).toBe('echo "Hi!"');
      expect(step1.workingDirectory).toBe(ctx.workingDirectory);
      expect(step1.shell).toBe(getDefaultShell());

      // - run:
      //     name: Say HELLO
      //     command: |
      //       echo "H"
      //       echo "E"
      //       echo "L"
      //       echo "L"
      //       echo "O"
      const step2 = buildSteps[1];
      expect(step2.id).toMatch(UUID_REGEX);
      expect(step2.name).toBe('Say HELLO');
      expect(step2.command).toMatchSnapshot();
      expect(step2.workingDirectory).toBe(ctx.workingDirectory);
      expect(step2.shell).toBe(getDefaultShell());

      // - run:
      //     id: id_2137
      //     command: echo "Step with an ID"
      const step3 = buildSteps[2];
      expect(step3.id).toBe('id_2137');
      expect(step3.name).toBeUndefined();
      expect(step3.command).toBe('echo "Step with an ID"');
      expect(step3.workingDirectory).toBe(ctx.workingDirectory);
      expect(step3.shell).toBe(getDefaultShell());

      // - run:
      //     name: List files
      //     working_directory: relative/path/to/files
      //     command: ls -la
      const step4 = buildSteps[3];
      expect(step4.id).toMatch(UUID_REGEX);
      expect(step4.name).toBe('List files');
      expect(step4.command).toBe('ls -la');
      expect(step4.workingDirectory).toBe(
        path.join(ctx.workingDirectory, 'relative/path/to/files')
      );
      expect(step4.shell).toBe(getDefaultShell());

      // - run:
      //     name: List files in another directory
      //     working_directory: /home/dsokal
      //     command: ls -la
      const step5 = buildSteps[4];
      expect(step5.id).toMatch(UUID_REGEX);
      expect(step5.name).toBe('List files in another directory');
      expect(step5.command).toBe('ls -la');
      expect(step5.workingDirectory).toBe('/home/dsokal');
      expect(step5.shell).toBe(getDefaultShell());

      // - run:
      //     name: Use non-default shell
      //     shell: /nib/hsab
      //     command: echo 123
      const step6 = buildSteps[5];
      expect(step6.id).toMatch(UUID_REGEX);
      expect(step6.name).toBe('Use non-default shell');
      expect(step6.command).toBe('echo 123');
      expect(step6.workingDirectory).toBe(ctx.workingDirectory);
      expect(step6.shell).toBe('/nib/hsab');
    });

    it('parses inputs', async () => {
      const ctx = createMockContext();
      const parser = new BuildConfigParser(ctx, {
        configPath: path.join(__dirname, './fixtures/inputs.yml'),
      });
      const workflow = await parser.parseAsync();
      const buildSteps = workflow.buildSteps;
      expect(buildSteps.length).toBe(1);

      // - run:
      //     name: Say HI
      //     inputs:
      //       name: Dominik Sokal
      //       country: Poland
      //     command: |
      //       echo "Hi, ${ inputs.name }!"
      const step1 = buildSteps[0];
      expect(step1.id).toMatch(UUID_REGEX);
      expect(step1.name).toBe('Say HI');
      expect(step1.command).toBe('echo "Hi, ${ inputs.name }!"');
      expect(step1.workingDirectory).toBe(ctx.workingDirectory);
      expect(step1.shell).toBe(getDefaultShell());
      expect(step1.inputs).toBeDefined();
      expect(step1.inputs?.[0].id).toBe('name');
      expect(step1.inputs?.[0].value).toBe('Dominik Sokal');
      expect(step1.inputs?.[1].id).toBe('country');
      expect(step1.inputs?.[1].value).toBe('Poland');
    });

    it('parses outputs', async () => {
      const ctx = createMockContext();
      const parser = new BuildConfigParser(ctx, {
        configPath: path.join(__dirname, './fixtures/outputs.yml'),
      });
      const workflow = await parser.parseAsync();
      const buildSteps = workflow.buildSteps;
      expect(buildSteps.length).toBe(2);

      // - run:
      //     outputs: [first_name, last_name]
      //     command: |
      //       set-output first_name Brent
      //       set-output last_name Vatne
      const step1 = buildSteps[0];
      expect(step1.id).toMatch(UUID_REGEX);
      expect(step1.name).toBeUndefined();
      expect(step1.command).toMatchSnapshot();
      expect(step1.workingDirectory).toBe(ctx.workingDirectory);
      expect(step1.shell).toBe(getDefaultShell());
      expect(step1.outputs).toBeDefined();
      expect(step1.outputs?.[0].id).toBe('first_name');
      expect(step1.outputs?.[0].required).toBe(true);
      expect(step1.outputs?.[1].id).toBe('last_name');
      expect(step1.outputs?.[1].required).toBe(true);

      // - run:
      //     outputs:
      //       - name: first_name
      //         required: true
      //       - name: middle_name
      //         required: false
      //       - name: last_name
      //       - nickname
      //     command: |
      //       set-output first_name Dominik
      //       set-output last_name Sokal
      //       set-output nickname dsokal
      const step2 = buildSteps[1];
      expect(step2.id).toMatch(UUID_REGEX);
      expect(step2.name).toBeUndefined();
      expect(step2.command).toMatchSnapshot();
      expect(step2.workingDirectory).toBe(ctx.workingDirectory);
      expect(step2.shell).toBe(getDefaultShell());
      expect(step2.outputs).toBeDefined();
      expect(step2.outputs?.[0].id).toBe('first_name');
      expect(step2.outputs?.[0].required).toBe(true);
      expect(step2.outputs?.[1].id).toBe('middle_name');
      expect(step2.outputs?.[1].required).toBe(false);
      expect(step2.outputs?.[2].id).toBe('last_name');
      expect(step2.outputs?.[2].required).toBe(true);
      expect(step2.outputs?.[3].id).toBe('nickname');
      expect(step2.outputs?.[3].required).toBe(true);
    });
  });
});
