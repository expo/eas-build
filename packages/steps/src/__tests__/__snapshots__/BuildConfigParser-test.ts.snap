// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BuildConfigParser parseAsync parses outputs 1`] = `
"set-output first_name Brent
set-output last_name Vatne
"
`;

exports[`BuildConfigParser parseAsync parses outputs 2`] = `
"set-output first_name Dominik
set-output last_name Sokal
set-output nickname dsokal
"
`;

exports[`BuildConfigParser parseAsync parses steps from the build workflow 1`] = `
"echo "H"
echo "E"
echo "L"
echo "L"
echo "O"
"
`;
